client/asset/eth/eth.go

```diff
@@ -313,28 +318,151 @@ func (*ExchangeWallet) PreRedeem(req *asset.PreRedeemForm) (*asset.PreRedeem, er
-// no redeem script.
-func (*ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
-	return nil, nil, asset.ErrNotImplemented
+// In UTXO based coins, the returned []dex.Bytes contains the redeem scripts for the
+// selected coins, but since there are no redeem scripts in Ethereum, the coin ids
+// are returned there.
+// Equal number of coins and redeem scripts must be returned.
+func (eth *ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
+	maxFees := ord.DEXConfig.MaxFeeRate * ord.DEXConfig.SwapSize * ord.MaxSwapCount
+	fundsNeeded := ord.Value + maxFees
+	balance, err := eth.Balance()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if balance.Available < fundsNeeded {
+		return nil, nil, fmt.Errorf("%v gwei are needed for order, but address %v only has %d available",
+			fundsNeeded, eth.acct.Address.Hex(), balance.Available)
+	}
+
+	coin := EthCoin{
+		address: eth.acct.Address,
+		value:   fundsNeeded,
+	}
+
+	eth.lockFunds(fundsNeeded)
+	return asset.Coins{coin}, []dex.Bytes{coin.ID()}, nil
 }
 
 // ReturnCoins unlocks coins. This would be necessary in the case of a
 // canceled order.
-func (*ExchangeWallet) ReturnCoins(unspents asset.Coins) error {
-	return asset.ErrNotImplemented
+func (eth *ExchangeWallet) ReturnCoins(unspents asset.Coins) error {
+	fundsToUnlock := make(map[string]uint64)
+	for _, coin := range unspents {
+		ethCoin, err := ethCoinFromID(coin.ID())
+		if err != nil {
+			return err
+		}
+
+		address := eth.acct.Address.Hex()
+		fundsToUnlock[address] += ethCoin.value
+		if fundsToUnlock[address] > eth.getLockedFunds(address) {
+			return fmt.Errorf("trying to return more coins than were taken from %v", address)
+		}
+	}
+
+	for address, amount := range fundsToUnlock {
+		eth.unlockFunds(address, amount)
+	}
+
+	return nil
 }
 
 // FundingCoins gets funding coins for the coin IDs. The coins are locked. This
 // method might be called to reinitialize an order from data stored externally.
-// This method will only return funding coins, e.g. unspent transaction outputs.
-func (*ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
-	return nil, asset.ErrNotImplemented
+func (eth *ExchangeWallet) FundingCoins(ids []dex.Bytes) (asset.Coins, error) {
+	balance, err := eth.Balance()
+	if err != nil {
+		return nil, err
+	}
+
+	var amountFunded uint64
+	coins := make([]asset.Coin, 0, len(ids))
+	for _, id := range ids {
+		ethCoin, err := ethCoinFromID(id)
+		if err != nil {
+			return nil, err
+		}
+		if ethCoin.address.Hex() != eth.acct.Address.Hex() {
+			return nil, fmt.Errorf("account: %v != coin's account: %v",
+				eth.acct.Address.Hex(), ethCoin.address.Hex())
+		}
+
+		amountFunded += ethCoin.value
+		coins = append(coins, ethCoin)
+	}
+
+	if amountFunded > balance.Available {
+		return nil, fmt.Errorf("value of FundingCoins: %v > available funds %v",
+			amountFunded, balance.Available)
+	}
+
+	eth.lockFunds(amountFunded)
+	return coins, nil
```

Comment:
I think you should also make sure there is only ever one processes getting funding coins at a time as you need to synchronize the locked funds. Two of this same method could get the same starting balance and end up locking more that is available.

Reply:
Good call.